### PR TITLE
refactor: pass vertices instead of transaction list to State Manager

### PR DESCRIPTION
**Pull request by dhedey**
Working with @malexandru-rdx 

This is still a work in progress :)

### CODE REVIEW

File: core-rust/state-manager/src/jni/state_computer.rs [hunk 1]
```diff
@@ -263,9 +264,10 @@ impl From<JavaCommitRequest> for CommitRequest {
     }
 }
 
-#[derive(Debug, Decode, Encode, Categorize)]
+#[derive(Debug, ScryptoCategorize, ScryptoEncode, ScryptoDecode)]
```
Comment (dhedey) [inline core-rust/state-manager/src/jni/state_computer.rs:267]: Out of interest, why did you have to change this to `ScryptoX`? Is there something that is only in the Scrypto model embedded in these types?

File: core/src/main/java/com/radixdlt/ledger/StateComputerLedger.java [hunk 2]
```diff
@@ -228,24 +226,56 @@ private Optional<ExecutedVertex> prepareInternal(
                 timeSupplier.currentTime()));
       }
 
-      final var executedTransactionsOptional =
-          this.verifier.verifyAndGetExtension(
+      // It's possibly that this function is called with a list of vertices which starts with some
```
Comment (dhedey) [inline core/src/main/java/com/radixdlt/ledger/StateComputerLedger.java:229]: It's possible*

**Review by dhedey — COMMENTED** [review]
Looks good, thanks @malexandru-rdx  - would appreciate a look over from @LukasGasior1  before merging 👍

File: core/src/main/java/com/radixdlt/ledger/StateComputerLedger.java [hunk 3]
```diff
@@ -228,24 +226,56 @@ private Optional<ExecutedVertex> prepareInternal(
...
+        if (committedAccumulatorHasMatchedStartOfAVertex) {
+          verticesInExtension.add(previousVertex);
+        }
+      }
+
+      if (!committedAccumulatorHasMatchedStartOfAVertex) {
+        // This could trigger if the "previous" vertices don't line up with the committed state.
+        // In other words, they don't provide a valid partial path from the committed state to the
+        // start of the proposal.
+        return Optional.empty();
```
Comment (LukasGasior1) [inline core/src/main/java/com/radixdlt/ledger/StateComputerLedger.java:258]: might be useful to add logs and/or metrics alongside Optional.empty() results

**Review by LukasGasior1 — APPROVED** [review]
just some renaming suggestions :)

File: core-rust/state-manager/src/jni/state_computer.rs (hunk 1 above)
Comment (malexandru-rdx) [reply core-rust/state-manager/src/jni/state_computer.rs]: `PreviousVertex` was using `ScryptoCategorize` and hence forced `PrepareRequest` to have the same.
It's kind of confusing what exactly each of the versions does so I just went with the flow :).

I've changed them to use `Decode`/`Encode`.

File: core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/PrepareRequest.java [hunk 4]
```diff
@@ -67,13 +67,15 @@
...
 import com.radixdlt.transactions.RawNotarizedTransaction;
 import com.radixdlt.utils.UInt64;
+import java.util.Arrays;
 import java.util.List;
+import java.util.Objects;
 
 public record PrepareRequest(
-    List<RawLedgerTransaction> previous,
+    byte[] parentAccumulator,
+    List<PreviousVertex> previous,
```
[the remark this answers is not in the document]
Comment (malexandru-rdx) [reply core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/PrepareRequest.java]: As of right now `prepare` function should support just a partial path up to root as long as there was a previous `prepare` that was up to the root from one of the vertices. I think it should work with no previous vertices as long as the proper `parentAccumulatorHash` is provided.

I would keep it as `previousVertices`.

File: core-rust/state-manager/src/jni/state_computer.rs [hunk 5]
```diff
@@ -263,9 +264,10 @@ impl From<JavaCommitRequest> for CommitRequest {
...
 pub struct JavaPrepareRequest {
-    pub already_prepared: Vec<JavaRawTransaction>,
+    pub parent_accumulator: Vec<u8>,
+    pub prepared_vertices: Vec<JavaPreviousVertex>,
```
[the remark this answers is not in the document]
Comment (malexandru-rdx) [reply core-rust/state-manager/src/jni/state_computer.rs]: Done

File: core-rust/state-manager/src/jni/mempool.rs [hunk 6]
```diff
@@ -202,27 +202,27 @@ fn do_get_transactions_to_relay(
 // DTO Models + Mapping
 //
 
-/// Corresponds to the payload_hash
+/// Corresponds to HashCode from Java
 #[derive(Debug, PartialEq, Eq, Categorize, Encode, Decode)]
-pub struct JavaPayloadHash(Vec<u8>);
+pub struct JavaHashCode(pub Vec<u8>);
```
Comment (malexandru-rdx) [inline core-rust/state-manager/src/jni/mempool.rs:207]: Is this fine? In Java it's represented as `HashCode`, it feels better to have the same naming.

**Pull request merged**